diff --git a/drivers/net/can/slcan.c b/drivers/net/can/slcan.c
index 3c28d1f..e7ad6a0 100644
--- a/drivers/net/can/slcan.c
+++ b/drivers/net/can/slcan.c
@@ -659,6 +659,8 @@ static int slcan_ioctl(struct tty_struct *tty, struct file *file,
 {
 	struct slcan *sl = (struct slcan *) tty->disc_data;
 	unsigned int tmp;
+	char cmd_buf[20];
+	int cmd_buf_size;
 
 	/* First make sure we're connected. */
 	if (!sl || sl->magic != SLCAN_MAGIC)
@@ -671,6 +673,25 @@ static int slcan_ioctl(struct tty_struct *tty, struct file *file,
 			return -EFAULT;
 		return 0;
 
+	case 22:
+		printk("slcan: changing bitrate\n");
+
+		/* invalid bitrate setting */
+		if (arg > 8)
+		{
+			printk("slcan: invalid bitrate set: %lu\n", arg);
+			return -EINVAL;
+		}
+
+		/* send commands to device: close -> set -> open */
+		spin_lock_bh(&sl->lock);
+		cmd_buf_size = sprintf(cmd_buf, "C\rS%lu\rO\r", arg);
+		set_bit(TTY_DO_WRITE_WAKEUP, &sl->tty->flags);
+		sl->tty->ops->write(sl->tty, cmd_buf, cmd_buf_size);
+		clear_bit(TTY_DO_WRITE_WAKEUP, &sl->tty->flags);
+		spin_unlock_bh(&sl->lock);
+		return 0;
+
 	case SIOCSIFHWADDR:
 		return -EINVAL;
 
